README.md
![CVA](/.github/assets/meta.png)

<h1 align="center">cva for elixir</h1>

<p align="center">Class Variance Authority for Elixir - easily construct classes with variant definitions.</p>

<p align="center">
  <a href="https://hexdocs.pm/cva">
    <img alt="Documentation" src="https://img.shields.io/badge/documentation-gray" />
  </a>
  <a href="https://hex.pm/packages/cva">
    <img alt="hex" src="https://img.shields.io/hexpm/v/cva.svg" />
  </a>
  <a href="https://twitter.com/benvp_">
    <img alt="Follow @benvp_ on Twitter" src="https://img.shields.io/twitter/follow/benvp_.svg?style=social&label=Follow" />
  </a>
</p>


## Introduction

Building out core HEEx function components like `button`, `heading`, `link` in general requires
some way to distinguish between different appearances of the component. This is generally achieved
by concatenating class strings or by extracting them into separate functions. In addition, maintaining proper definitions of supported `attr` `:value` options is required.

`ex_cva` aims to make this process convenient by providing a clean and maintainable way to define component variants.

## Installation

Add the `cva` dependency to your `mix.exs` file.

```elixir
def deps do
  [
    {:cva, "~> 0.2"}
  ]
end
```

## Usage with HEEx function components

Configure a few variants with defaults and optionally add compound variants. CVA will take care of creating the proper class names.

One more goodie: it even creates compile-time checks for your variants to make sure all required
variants are applied and contain correct values (thanks to `Phoenix.Component.attr/3`).

```elixir
defmodule MyWeb.Components do
  use Phoenix.Component
  use CVA.Component

  variant :intent, [
      primary: "bg-cyan-600",
      secondary: "bg-zinc-700",
      destructive: "bg-red-500"
    ],
    default: :secondary

  variant :size, [
      xs: "rounded px-2.5 py-1.5 text-xs",
      sm: "rounded-md px-3 py-2 text-sm",
      md: "rounded-md px-4 py-2 text-sm",
      lg: "rounded-md px-4 py-2 text-base",
      xl: "rounded-md px-6 py-3 text-base"
    ],
    default: :md

  compound_variant "uppercase", intent: :primary, size: :xl

  attr :rest, :global
  slot :inner_block

  def button(assigns) do
    ~H"""
    <button class={@cva_class} {@rest}><%= render_slot(@inner_block) %></button>
    """
  end
end

defmodule MyWeb.SomeLive do
  import MyWeb.Components

  def render(assigns) do
    ~H"""
    <.button intent="primary">Click me</.button>
    """
  end
end
```

## Raw `cva` usage

Even though `ex_cva` shines when working with function components, you can still use the raw `cva` function to generate classes.

```elixir
defmodule MyCVA do
  import CVA

  def button(props) do
    config = [
      variants: [
        intent: [
          primary: "bg-cyan-600",
          secondary: "bg-zinc-700",
          destructive: "bg-red-500"
        ],
        size: [
          xs: "rounded px-2.5 py-1.5 text-xs",
          sm: "rounded-md px-3 py-2 text-sm",
          md: "rounded-md px-4 py-2 text-sm",
          lg: "rounded-md px-4 py-2 text-base",
          xl: "rounded-md px-6 py-3 text-base"
        ]
      ]
    ]

    cva(config, props)
  end
end

button(intent: :primary, size: :md) # -> "bg-cyan-600 rounded-md px-4 py-2 text-sm"
```

## Acknowledgements

- [**cva**](https://github.com/joe-bell/cva) ([Joe Bell](https://github.com/joe-bell))
  Thank you for providing the JavaScript implementation. It inspired me to port this over to Elixir. I hope you don't mind that I hijacked your logo. If you are not feeling well about that, feel free to shoot me a message.

## Contributing

Contributions are very welcome. To get it up on your local machine, just check out the repo and run

```bash
mix deps.get
mix test
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [v0.2.0] (2022-04-06)

## Improvements

* Add `.formatter.exs` to files to allow `import_deps: [:cva]`.

## [v0.2.0] (2022-12-06)

### Improvements

* `variant/2` now properly supports boolean values. You can now do something like the following:

```elixir
variant :disabled,
        [true: "disabled-class", false: "enabled-class"],
        default: false

# or

variant :disabled,
        [true: "disabled-class"],
        default: nil

def button(assigns) do
  ~H"""
  <button class={@cva_class} disabled={@disabled}>
    <%= render_slot(@inner_block) %>
  </button>
  """
end

# ... where you use that component

<.button disabled>Click me</.button>

# -> <button class="disabled-class" disabled>Click me</button>
```

### Changes

* `variant/2` does not automatically infer the `required` option anymore. If you want to make a variant mandatory, you have to provide the `required: true` option.

* `variant/2` now allows `nil` as a valid value.

### Bugfixes

* Fixes an issue where defining 3 variants would cause an `attr` already defined error.


## [v0.1.2] (2022-12-03)

### Bugfixes

* Add `variant: 2` to formatter.
* Fixes an issue where a `class` assign would accidentally be merged into `@cva_class` without
  explicitly declaring it.

## [v0.1.1] (2022-11-26)

### Bugfixes

* Fix formatter exports

## [v0.1.0] (2022-11-26)

🚀 Initial release


test/cva_test.exs
defmodule CVATest do
  use ExUnit.Case

  import CVA

  describe "cx" do
    test "correctly merges classes 1" do
      class = cx([nil, ""])
      assert class == ""
    end

    test "correctly merges classes 2" do
      class = cx([["foo", nil, "bar", "baz"]])
      assert class == "foo bar baz"
    end

    test "correctly merges classes 3" do
      class = cx(["foo", nil, "bar", "baz"])
      assert class == "foo bar baz"
    end

    test "correctly merges classes 4" do
      class =
        cx([
          "foo",
          [
            nil,
            ["bar"],
            [
              [
                "baz",
                "qux",
                "quux",
                "quuz",
                [[[[[[[[["corge", "grault"]]]]], "garply"]]]]
              ]
            ]
          ]
        ])

      assert class == "foo bar baz qux quux quuz corge grault garply"
    end
  end
end


mix.exs
defmodule CVA.MixProject do
  use Mix.Project

  @version "0.2.1"

  def project do
    [
      app: :cva,
      version: @version,
      elixir: "~> 1.13",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      package: package(),
      docs: docs(),
      name: "CVA",
      homepage_url: "https://github.com/benvp/ex_cva",
      description: """
      Class Variance Authority.
      Easily construct classes with variant definitions.
      """
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:phoenix_live_view, "~> 0.18"},
      {:ex_doc, "~> 0.29", only: :dev, runtime: false}
    ]
  end

  defp docs do
    [
      main: "CVA",
      source_ref: "v#{@version}",
      source_url: "https://github.com/benvp/ex_cva"
    ]
  end

  defp package do
    [
      maintainers: ["Benjamin von Polheim"],
      licenses: ["MIT"],
      links: %{
        Changelog: "https://hexdocs.pm/cva/changelog.html",
        GitHub: "https://github.com/benvp/ex_cva"
      },
      files:
        ~w(lib) ++
          ~w(CHANGELOG.md LICENSE.md mix.exs README.md .formatter.exs)
    ]
  end
end


lib/cva/component.ex
defmodule CVA.Component do
  @moduledoc ~S'''
  Integrates CVA with Phoenix function components.

  The `variant/3` and `compound_variant/2` macros allow easy definition of variants and compound
  variants for a component. This also includes compile time checks for the specified variants.

  When using this module, please make sure that you add include :cva to your imports in
  `.formatter.exs`.

  ```elixir
  [
    import_deps: [:cva],
  ]
  ```

  ## Usage

      defmodule MyWeb.Components do
        use CVA.Component

        variant :intent, [
            primary: "bg-cyan-600",
            secondary: "bg-zinc-700",
            destructive: "bg-red-500"
          ],
          default: :secondary

        variant :size, [
            xs: "rounded px-2.5 py-1.5 text-xs",
            sm: "rounded-md px-3 py-2 text-sm",
            md: "rounded-md px-4 py-2 text-sm",
            lg: "rounded-md px-4 py-2 text-base",
            xl: "rounded-md px-6 py-3 text-base"
          ],
          default: :md

        compound_variant "uppercase", intent: :primary, size: :xl

        attr :rest, :global
        slot :inner_block

        def button(assigns) do
          ~H"""
          <button class={@cva_class} {@rest}><%= render_slot(@inner_block) %></button>
          """
        end
      end

      defmodule MyWeb.SomeLive do
        import MyWeb.Components

        def render(assigns) do
          ~H"""
          <.button intent="primary">Click me</.button>
          """
        end
      end
  '''

  import CVA
  import Phoenix.Component

  @doc ~S'''
  Declares a variant for HEEx function components.

  When declaring variants, an assign `cva_class` is added to the component. This assign contains
  the class according to all variant definitions of the component. This class is intended to
  be passed into the `class` attribute of the component.

  A component can have multiple variants.

  ## Arguments

    * `name` - an atom defining the name of the attribute. Note that attributes cannot define the
    same name as any other attributes or slots or attributes declared for the same component.

    * `variants` - a keyword list of variants. Supported variant values are strings and a list of strings.

    * `opts` - a keyword list of options. Defaults to `[]`.

  ## Options

      * `default` - the default variant.

      All other options will be passed to `Phoenix.Component.attr/3`.

  ## Example

      defmodule MyWeb.Components do
        use CVA.Component

        variant :intent, [
            primary: "bg-cyan-600",
            secondary: "bg-zinc-700",
            destructive: "bg-red-500"
          ],
          default: :secondary

        variant :size, [
            xs: "rounded px-2.5 py-1.5 text-xs",
            sm: "rounded-md px-3 py-2 text-sm",
            md: "rounded-md px-4 py-2 text-sm",
            lg: "rounded-md px-4 py-2 text-base",
            xl: "rounded-md px-6 py-3 text-base"
          ],
          default: :md

        def button(assigns) do
          ~H"""
          <button class={@cva_class} {@rest}><%= render_slot(@inner_block) %></button>
          """
        end
      end
  '''
  defmacro variant(name, variants, opts \\ [])
           when is_atom(name) and is_list(variants) and is_list(opts) do
    quote do
      unless Module.get_attribute(__MODULE__, :__cva_variant_called__) do
        attr(:cva_class, :string, default: nil)
      end

      cva_variant = %{
        name: unquote(name),
        variants: unquote(variants),
        default: unquote(opts[:default]),
        line: __ENV__.line,
        file: __ENV__.file
      }

      Module.put_attribute(__MODULE__, :__cva_variants__, cva_variant)
      Module.put_attribute(__MODULE__, :__cva_variant_called__, true)

      attr(unquote_splicing(CVA.Component.__attr__!({name, variants}, opts)))
    end
  end

  @doc ~S'''
  Declares a compound variant for HEEx function components.

  A compound variant defines a set of required variants. If all variants are present, the
  given class will be assigned.

  A component can have multiple compound variants.

  ## Arguments

    * `class` - the class to add if all variants are present.

    * `variants` - a keyword list of required variants.

  ## Example

      defmodule MyWeb.Components do
        use CVA.Component

        variant :intent, [
            primary: "bg-cyan-600",
            secondary: "bg-zinc-700",
          ],
          default: :secondary

        variant :size, [
            md: "rounded-md px-4 py-2 text-sm",
            xl: "rounded-md px-6 py-3 text-base"
          ],
          default: :md

        compound_variant "uppercase", intent: :primary, size: :xl

        def button(assigns) do
          ~H"""
          <button class={@cva_class} {@rest}><%= render_slot(@inner_block) %></button>
          """
        end
      end
  '''
  defmacro compound_variant(class, variants) do
    quote do
      @__cva_compound_variants__ %{
        variants: unquote(variants),
        class: unquote(class)
      }
    end
  end

  defp pop_variants(env) do
    variants = Module.delete_attribute(env.module, :__cva_variants__) || []
    Enum.reverse(variants)
  end

  defp pop_compound_variants(env) do
    variants = Module.delete_attribute(env.module, :__cva_compound_variants__) || []
    Enum.reverse(variants)
  end

  def __attr__!({name, variants}, opts) do
    values =
      variants
      |> Keyword.keys()
      |> Enum.map(fn
        v when is_boolean(v) -> v
        v -> Atom.to_string(v)
      end)

    cva_opts = [values: values ++ [nil]]

    cva_opts =
      cond do
        is_boolean(opts[:default]) -> Keyword.put(cva_opts, :default, opts[:default])
        opts[:default] != nil -> Keyword.put(cva_opts, :default, Atom.to_string(opts[:default]))
        true -> cva_opts
      end

    opts = Keyword.merge(opts, cva_opts)

    type =
      cond do
        Enum.all?(values, &is_boolean/1) -> :boolean
        Enum.all?(values, &is_binary/1) -> :string
        true -> :any
      end

    [name, type, opts]
  end

  def __on_definition__(env, kind, name, _args, _guards, _body) do
    variant_defs = pop_variants(env)
    compound_variant_defs = pop_compound_variants(env)

    Module.put_attribute(env.module, :__cva_variant_called__, false)

    if not String.starts_with?(to_string(name), "__") and variant_defs != [] do
      default_variants =
        for %{name: name, default: default} <- variant_defs,
            not is_nil(default),
            do: {name, default}

      variants = for %{name: name, variants: v} <- variant_defs, do: {name, v}

      compound_variants =
        for %{variants: variants, class: class} <- compound_variant_defs,
            do: Keyword.put(variants, :class, class)

      configs =
        env.module
        |> Module.get_attribute(:__cva__)
        |> Map.put(name, %{
          kind: kind,
          config: [
            variants: variants,
            default_variants: default_variants,
            compound_variants: compound_variants
          ]
        })

      Module.put_attribute(env.module, :__cva__, configs)
    end

    :ok
  end

  defmacro __before_compile__(env) do
    configs = Module.get_attribute(env.module, :__cva__)

    names_and_defs =
      for {name, %{kind: kind, config: config}} <- configs do
        variant_names = for {name, _} <- config[:variants], do: name

        body =
          quote do
            cva_props = Map.take(assigns, unquote(variant_names))

            assigns =
              Phoenix.Component.assign(
                assigns,
                :cva_class,
                cva(unquote(config), cva_props)
              )

            super(assigns)
          end

        full_def =
          quote do
            unquote(kind)(unquote(name)(assigns)) do
              unquote(body)
            end
          end

        {{name, 1}, full_def}
      end

    {names, defs} = Enum.unzip(names_and_defs)

    overridable =
      if names != [] do
        quote do
          defoverridable unquote(names)
        end
      end

    {:__block__, [], [overridable | defs]}
  end

  defmacro __using__(_opts \\ []) do
    quote do
      import CVA
      import unquote(__MODULE__)

      Module.put_attribute(__MODULE__, :before_compile, unquote(__MODULE__))
      Module.put_attribute(__MODULE__, :on_definition, unquote(__MODULE__))
      Module.put_attribute(__MODULE__, :__cva__, %{})
      Module.put_attribute(__MODULE__, :__cva_variant_called__, false)

      Module.register_attribute(__MODULE__, :__cva_variants__, accumulate: true)
      Module.register_attribute(__MODULE__, :__cva_compound_variants__, accumulate: true)
    end
  end
end


lib/cva.ex
defmodule CVA do
  @moduledoc """
  Construct classes with variant definitions.

  A variant consists of a name and a class. These variants are defined by providing a declarative
  set of nested keyword lists.

      defmodule MyCVA do
        import CVA

        def button(props) do
          config = [
            variants: [
              intent: [
                primary: "bg-cyan-600",
                secondary: "bg-zinc-700",
                destructive: "bg-red-500"
              ],
              size: [
                xs: "rounded px-2.5 py-1.5 text-xs",
                sm: "rounded-md px-3 py-2 text-sm",
                md: "rounded-md px-4 py-2 text-sm",
                lg: "rounded-md px-4 py-2 text-base",
                xl: "rounded-md px-6 py-3 text-base"
              ]
            ]
          ]

          cva(config, props)
        end
      end

    Invoking `button/1` with the desired variants will return a class name including
    the classes specified in the config.

        button(intent: :primary, size: :md) # -> "bg-cyan-600 rounded-md px-4 py-2 text-sm"

    See `cva/3` for more info.


    ## Usage with Phoenix function components.

    CVA integrates nicely with Phoenix function components by providing `variant/3` and
    `compound_variant/2` macros.

    Please refer to the `CVA.Component` module.
  """

  @doc """
  Merges a list of classes into a single class string removing any nil or empty strings.
  Class lists can be infinitely nested.
  """
  def cx(classes) when is_binary(classes), do: classes

  def cx(classes) when is_list(classes) do
    classes
    |> List.flatten()
    |> Enum.filter(&(!!&1 && &1 != ""))
    |> Enum.join(" ")
  end

  @doc """
  See `cva/3`.
  """
  def cva(config) when is_list(config), do: cva("", config, [])

  @doc """
  See `cva/3`.
  """
  def cva(config, props), do: cva("", config, props)

  def cva(base, config, props) when is_list(config) do
    cva(
      base,
      Enum.into(config, %{variants: nil, compound_variants: nil, default_variants: nil}),
      Enum.into(props, %{})
    )
  end

  def cva(base, %{variants: nil}, props), do: cx([base, props[:class]])

  @doc """
  Construct a class string from a variant configuration.

  Accepts a base class string, a variant configuration, and a list of props.

      config = [
        variants: [
          intent: [
            primary: "bg-cyan-600",
            secondary: "bg-zinc-700",
            destructive: "bg-red-500"
          ],
          size: [
            xs: "rounded px-2.5 py-1.5 text-xs",
            sm: "rounded-md px-3 py-2 text-sm",
            md: "rounded-md px-4 py-2 text-sm",
            lg: "rounded-md px-4 py-2 text-base",
            xl: "rounded-md px-6 py-3 text-base"
          ]
        ],
        default_variants: [
          intent: :secondary,
          size: :md
        ],
        compound_variants: [
          [intent: :primary, size: :xl, class: "uppercase"]
        ]
      ]

      cva("base", config, intent: :primary, size: :md) # -> "base bg-cyan-600 rounded-md px-4 py-2 text-sm"
      cva("base", config, intent: :primary, size: :xl) # -> "base bg-cyan-600 rounded-md px-6 py-3 text-base uppercase"
      cva("base", config, intent: :primary, size: :xl) # -> "base bg-cyan-600 rounded-md px-6 py-3 text-base uppercase"

      cva("base", config) # -> "base bg-zinc-700 rounded-md px-4 py-2 text-sm"
      cva("base", config, intent: :primary) # -> "base bg-cyan-600 rounded-md px-4 py-2 text-sm"

  ## Config

  The configuration is a keyword list with the following keys:

    * `:variants` - A keyword list of variants. Each variant is a keyword list of
      variant names and classes.
    * `:compound_variants` - A list of compound variants. Each compound
      variant is a keyword list of required variants and a `:class` key to specify the class
      which should be applied if all variants are present.
    * `:default_variants` - A keyword list of default variants. For example `[intent: :primary]`.

  ## Props

  Props define the variants to be applied. Each key in the props list must be a variant name.
  Values can either be an atom or a string.

  ### Special props

    * `:class` - A class string or list of classes. This class is applied last.
  """
  def cva(base, config, props) do
    cx([
      base,
      variant_class_names(config, props),
      compound_variant_class_names(config, props),
      props[:class]
    ])
  end

  defp variant_class_names(_config, %{variants: nil}), do: []

  defp variant_class_names(config, props) do
    config[:variants]
    |> Keyword.keys()
    |> Enum.map(fn variant ->
      if Map.has_key?(props, variant) && props[variant] == nil do
        nil
      else
        variant_prop = props[variant]
        default_variant_prop = config[:default_variants][variant]

        variant_key =
          case variant_prop do
            p when is_binary(p) -> String.to_existing_atom(p)
            p -> p
          end || default_variant_prop

        # In case we receive nil, convert it to false to properly support
        # boolean variants. E.g. `disabled: [true: "i-am-disabled"]`
        variant_key = if variant_key, do: variant_key, else: false

        config[:variants][variant][variant_key]
      end
    end)
  end

  defp compound_variant_class_names(config, props) do
    compound_variants = config[:compound_variants]

    if compound_variants do
      compound_variants
      |> Enum.reduce([], fn compound_variant, acc ->
        {class, compound} = Keyword.pop(compound_variant, :class)
        props_with_default = Keyword.merge(config[:default_variants] || [], Map.to_list(props))

        match? =
          Enum.all?(compound, fn {k, v} ->
            p =
              if is_binary(props_with_default[k]),
                do: String.to_existing_atom(props_with_default[k]),
                else: props_with_default[k]

            p == v
          end)

        if match? do
          [class | acc]
        else
          acc
        end
      end)
    end
  end
end
